apps/web/src/hooks/useFormAction.ts

```diff
@@ -58,10 +62,14 @@ export function useFormAction<
 
   const action = useCallback(
     async (formData: FormData) => {
-      const json = formDataToJson(
+      let json = formDataToJson(
         formData,
       ) as inferServerActionInput<TServerAction>
 
+      if (onSubmit) {
+        json = onSubmit(json)
```

Comment:
hmm this smells like @andresgutgon recommendation.. Are we sure about this? a callback that needs to return the payload it receives and inadvertendly changes the payload schema that gets sent to the api... leaky abstraction

If you need onSubmit from form you could just have used the server action directly, example:

```typescript
const { update } = useProviderApiKeys()
const onSubmit = (e) => {
    const formData = new FormData(e.currentTarget)
    const json = formDataToJson(formData)

    update(json)
}

return <form onSubmit={onSubmit) ... />
```

same number of lines, no leaky abstraction involved